### DISCUSSION TITLE
Parent / Child Category management

### DISCUSSION THREAD

**Discussion by futurewebsites** [area17/twill#1148] [Q&A] [answered]
Hi Folks,

I have an existing category structure: 

Categories:
`id, parent_id, name, description, order`

The Category model has a recursive method to get all children of a parent and for the inverse the parent of a child the structure allows for a category tree with unlimited depth

```
public function parent()
    {
        return $this->belongsTo(Category::class, 'parent_id', 'id');
    }


    public function children()
    {
        return $this->hasMany(Category::class, 'parent_id', 'id');
    }

    // recursive, loads all children
    public function childrenRecursive()
    {
        return $this->children()->with('childrenRecursive')->orderBy('order');
    }

    // recursive, loads all parents
    public function parentRecursive($parent_id)
    {
        if ($cat = $this->select('name', 'parent_id')->where("id", $parent_id)->first()) {
            $this->paths[] = Str::slug($cat->name);

            if ($cat->parent_id > 0) {
                $this->parentRecursive($cat->parent_id);
            }
        }
    }
```

Firstly is there or has any created any sort of category tree management using twill? 

Our current PHP category management page shows the entire category tree and has drag and drop re-ordering this allows us to visualise the entire category structure. A parent category can be added via a button at the top of the index view and each parent or child row has CRUD links with the addition of and Add Child Category button

I need to replicate this or a close to using Twill any pointers in the right direction would be greatly appreciated.

**Comment by pboivin** [comment, marked as answer]
Hi @futurewebsites ,

not sure if this will cover 100% of your needs but it's possible to have nested modules in the index UI, using the [`kalnoy/nestedset`](https://packagist.org/packages/kalnoy/nestedset) package. From the docs: https://twill.io/docs/#nested-module

Quick example:

![2021-09-08_17-07](https://user-images.githubusercontent.com/7805679/132585800-95e4b98e-b8ce-475c-8f04-ecde064fa8e5.png)

**Reply by futurewebsites** [reply]
@pboivin thank you that looks on the right track :)
